honeybadgermpc/protocols/avid.py

```diff
@@ -0,0 +1,212 @@
+        where ``tag`` is one of ``{"VAL", "ECHO", "READY"}``
+        :param send: sends (without blocking) a message to a designated
+        recipient ``send(i, (tag, ...))``
+        """
+        assert n >= 3 * f + 1
+        assert f >= 0
+        assert 0 <= leader < n
+
+        self._data = [None] * n
+        self.n = n
+        self.f = f
+        self.leader = leader
+        self.recv = recv
+        self.send = send
+        # size of input_list for disperse
+        self.input_size = input_size
+        # the response for retrieve
+        self.retrieval_queue = asyncio.Queue()
+        # A blocking queue: retrive will be sent until the ok state is reached
+        self.ok_queue = asyncio.Queue()
+
+    async def retrieve(self, sid, index):
+        """ Retrieve individual item
+        :param int index:  the index of retrieval
+        """
+        # we are able to retrieve things from AVID indivdually
+        assert 0 <= index < self.input_size
+
+        # send retrieve until ok state
+        assert await self.ok_queue.get()
+
+        # send retrieve to all the parties
+        for i in range(self.n):
+            self.send(i, (sid, "RETRIEVE", index))
+
+        result = [None] * self.input_size
+        response_set = set()
+
+        # response threshold same as k
+        response_threshold = self.n - 2 * self.f
+
+        def decode_output(roothash):
+            m = decode(response_threshold, self.n, result)
+            # Rebuild the merkle tree to guarantee decoding is correct
+            _stripes = encode(response_threshold, self.n, m)
+            _mt = merkle_tree(_stripes)
+            _roothash = _mt[1]
+            assert _roothash == roothash
+            return m
+
+        while True:  # recv loop for retrieve
+            sender, msg = await self.retrieval_queue.get()
+            if msg[1] == 'RESPONSE':
+                (_, _, roothash, data) = msg
+
+                if sender in response_set:
+                    logger.warning("[{pid}] Redundant RESPONSE")
+                    continue
+
+                result[sender] = data
+                response_set.add(sender)
+
+                if len(response_set) >= response_threshold:
+                    m = decode_output(roothash)
+                    return m
+
+    async def disperse(self, sid, pid, input_list):
+        """ Main information dispersal handling
+        :param int sid: e.x. tag to be used
+        :param int pid: current member id
+        :param int input_list: the list of message
+            for each party from 1 to n from the dealer
+        """
+        # setup some parameters
+        k = self.n - 2 * self.f  # Wait to reconstruct.
+        echo_threshold = math.ceil((self.n + self.f + 1) / 2)  # number of ECHOs needed
+        ready_threshold = self.f + 1  # number of READYs needed for READY
+        output_threshold = 2 * self.f + 1  # number of READYs needed for storing
+
+        def broadcast(o):
```

Comment:
You can declare this in the constructor and then reuse this even in the `retrieve` method.

Reply:
Resolved.